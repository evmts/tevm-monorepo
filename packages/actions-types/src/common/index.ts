export type { Abi } from './Abi.js'
export type { Block } from './Block.js'
export type { BlockParam } from './BlockParam.js'
export type { BlockTag } from './BlockTag.js'
export type { Address } from './Address.js'
export type { Hex } from './Hex.js'
export type { EmptyParams } from './EmptyParams.js'
export type { Log } from './Log.js'
export type { BlockResult } from './BlockResult.js'
export type { FilterLog } from './FilterLog.js'
export type { FilterParams } from './FilterParams.js'
export type { TransactionParams } from './TransactionParams.js'
export type { TransactionResult } from './TransactionResult.js'
export type { TransactionReceiptResult } from './TransactionReceiptResult.js'
export type { TraceCall } from './TraceCall.js'
export type { TraceResult } from './TraceResult.js'
export type { TraceType } from './TraceType.js'
export type { NetworkConfig } from './NetworkConfig.js'
