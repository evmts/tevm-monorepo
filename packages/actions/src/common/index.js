export { isValidCallEvents, validateCallEvents } from './zCallEvents.js'
