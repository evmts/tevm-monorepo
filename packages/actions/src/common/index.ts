export type { Abi } from './Abi.js'
export type { AccountState } from './AccountState.js'
export type { Address } from './Address.js'
export type { Block } from './Block.js'
export type { BlockOverrideSet } from './BlockOverrideSet.js'
export type { BlockParam } from './BlockParam.js'
export type { BlockResult } from './BlockResult.js'
export type { BlockTag } from './BlockTag.js'
export type { CallEvents, Message, NewContractEvent } from './CallEvents.js'
export type { CallTraceResult } from './CallTraceResult.js'
export type { EmptyParams } from './EmptyParams.js'
export type { FlatCallTraceResult, FlatCallTraceCall } from './FlatCallTraceResult.js'
export type { FilterLog } from './FilterLog.js'
export type { FilterParams } from './FilterParams.js'
export type { Hex } from './Hex.js'
export type { Log } from './Log.js'
export type { NetworkConfig } from './NetworkConfig.js'
export type { PrestateTraceResult } from './PrestateTraceResult.js'
export type { StateOverrideSet } from './StateOverrideSet.js'
export type { StructLog } from './StructLog.js'
export type { TraceCall } from './TraceCall.js'
export type { TraceResult } from './TraceResult.js'
export type { TraceType } from './TraceType.js'
export type { TransactionParams } from './TransactionParams.js'
export type { TransactionReceiptResult } from './TransactionReceiptResult.js'
export type { TransactionResult } from './TransactionResult.js'
