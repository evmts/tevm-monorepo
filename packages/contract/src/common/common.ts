export { ERC2771Context, ERC2771Forwarder } from './metatx'
export { Multicall, Multicall3 } from './multicall'
export {
	ERC1155,
	ERC20,
	ERC20Votes,
	ERC2981,
	ERC721,
	ERC721Votes,
} from './tokens'
