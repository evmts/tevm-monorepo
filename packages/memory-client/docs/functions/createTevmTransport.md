[**@tevm/memory-client**](../README.md) • **Docs**

***

[@tevm/memory-client](../globals.md) / createTevmTransport

# Function: createTevmTransport()

> **createTevmTransport**(`options`): [`TevmTransport`](../type-aliases/TevmTransport.md)\<`string`\>

## Parameters

• **options**: `TevmNodeOptions`\<`Common`\> = `{}`

Configuration options for the base client, similar to those used in `memoryClient` or a low-level `baseClient`.

## Returns

[`TevmTransport`](../type-aliases/TevmTransport.md)\<`string`\>

A configured TEVM transport.

## Defined in

[packages/memory-client/src/createTevmTransport.js:58](https://github.com/evmts/tevm-monorepo/blob/main/packages/memory-client/src/createTevmTransport.js#L58)
