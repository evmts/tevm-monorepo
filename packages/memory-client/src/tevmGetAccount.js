import { getAccountHandler } from '@tevm/actions'

/**
 * A tree-shakeable version of the `tevmGetAccount` action for viem.
 * Retrieves detailed account information from the TEVM state, including balance, nonce, code, and optionally storage.
 *
 * This function provides a comprehensive view of an Ethereum account's state within the TEVM environment.
 * It returns account properties including:
 * - Address (the requested account address)
 * - Balance (in wei)
 * - Nonce (transaction count)
 * - Code (bytecode for contract accounts)
 * - Storage (optionally, contract storage slots and values)
 *
 * The `returnStorage` parameter controls whether the contract's storage is included in the response:
 * - When `false` (default): Only basic account information is returned, which is more efficient
 * - When `true`: Contract storage slots and values are also returned
 *
 * **Important fork mode considerations:**
 * In fork mode, the function only returns storage slots that have already been accessed and cached in the local state.
 * If a storage slot hasn't been previously accessed in your TEVM session, it won't appear in the results even with
 * `returnStorage: true`. To ensure all relevant storage is included, you may need to interact with the contract first.
 *
 * **Performance note:** Requesting storage for contracts with many storage slots can be computationally expensive
 * and produce large response objects. Use `returnStorage: true` judiciously for contracts with known limited storage.
 *
 * @param {import('viem').Client<import('./TevmTransport.js').TevmTransport<string>>} client - The viem client configured with TEVM transport.
 * @param {import('@tevm/actions').GetAccountParams} params - Parameters for retrieving the account information.
 * @returns {Promise<import('@tevm/actions').GetAccountResult>} The account information, including address, balance, nonce, code, and optionally storage.
 *
 * @example
 * ```typescript
 * import { tevmGetAccount } from 'tevm/actions'
 * import { createClient, http } from 'viem'
 * import { optimism } from 'tevm/common'
 * import { createTevmTransport } from 'tevm'
 *
 * const client = createClient({
 *   transport: createTevmTransport({
 *     fork: { transport: http('https://mainnet.optimism.io')({}) }
 *   }),
 *   chain: optimism,
 * })
 *
 * async function example() {
 *   // Get EOA (externally owned account) information
 *   const userAccount = await tevmGetAccount(client, {
 *     address: '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',
 *   })
 *   console.log(`Balance: ${userAccount.balance}`)
 *   console.log(`Nonce: ${userAccount.nonce}`)
 *
 *   // Get contract account with storage
 *   const contractAccount = await tevmGetAccount(client, {
 *     address: '0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984', // UNI token
 *     returnStorage: true,
 *   })
 *   console.log(`Contract bytecode: ${contractAccount.code}`)
 *   console.log(`Storage slots: ${Object.keys(contractAccount.storage || {}).length}`)
 *
 *   // Storage is a mapping of hex slot keys to hex values
 *   // Both are represented as hex strings
 *   if (contractAccount.storage) {
 *     for (const [slot, value] of Object.entries(contractAccount.storage)) {
 *       console.log(`Slot ${slot}: ${value}`)
 *     }
 *   }
 * }
 *
 * example()
 * ```
 *
 * @see [GetAccountParams](https://tevm.sh/reference/tevm/actions/type-aliases/getaccountparams/) for options reference.
 * @see [GetAccountResult](https://tevm.sh/reference/tevm/actions/type-aliases/getaccountresult/) for return values reference.
 * @see [TEVM Actions Guide](https://tevm.sh/learn/actions/)
 * @see [tevmSetAccount](https://tevm.sh/reference/tevm/actions/functions/setaccounthandler/) for modifying account state
 */
export const tevmGetAccount = async (client, params) => {
	return getAccountHandler(client.transport.tevm)(params)
}
