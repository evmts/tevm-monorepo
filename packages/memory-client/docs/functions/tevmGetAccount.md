[**@tevm/memory-client**](../README.md)

***

[@tevm/memory-client](../globals.md) / tevmGetAccount

# Function: tevmGetAccount()

> **tevmGetAccount**(`client`, `params`): `Promise`\<`GetAccountResult`\<`TevmGetAccountError`\>\>

Defined in: [packages/memory-client/src/tevmGetAccount.js:78](https://github.com/evmts/tevm-monorepo/blob/main/packages/memory-client/src/tevmGetAccount.js#L78)

A tree-shakeable version of the `tevmGetAccount` action for viem.
Retrieves detailed account information from the TEVM state, including balance, nonce, code, and optionally storage.

This function provides a comprehensive view of an Ethereum account's state within the TEVM environment.
It returns account properties including:
- Address (the requested account address)
- Balance (in wei)
- Nonce (transaction count)
- Code (bytecode for contract accounts)
- Storage (optionally, contract storage slots and values)

The `returnStorage` parameter controls whether the contract's storage is included in the response:
- When `false` (default): Only basic account information is returned, which is more efficient
- When `true`: Contract storage slots and values are also returned

**Important fork mode considerations:**
In fork mode, the function only returns storage slots that have already been accessed and cached in the local state.
If a storage slot hasn't been previously accessed in your TEVM session, it won't appear in the results even with
`returnStorage: true`. To ensure all relevant storage is included, you may need to interact with the contract first.

**Performance note:** Requesting storage for contracts with many storage slots can be computationally expensive
and produce large response objects. Use `returnStorage: true` judiciously for contracts with known limited storage.

## Parameters

### client

`Client`\<[`TevmTransport`](../type-aliases/TevmTransport.md)\<`string`\>, `undefined` \| `Chain`, `undefined` \| `Account`, `undefined`, `undefined` \| \{[`key`: `string`]: `unknown`; `account?`: `undefined`; `batch?`: `undefined`; `cacheTime?`: `undefined`; `ccipRead?`: `undefined`; `chain?`: `undefined`; `key?`: `undefined`; `name?`: `undefined`; `pollingInterval?`: `undefined`; `request?`: `undefined`; `transport?`: `undefined`; `type?`: `undefined`; `uid?`: `undefined`; \}\>

The viem client configured with TEVM transport.

### params

`GetAccountParams`\<`boolean`\>

Parameters for retrieving the account information.

## Returns

`Promise`\<`GetAccountResult`\<`TevmGetAccountError`\>\>

The account information, including address, balance, nonce, code, and optionally storage.

## Example

```typescript
import { tevmGetAccount } from 'tevm/actions'
import { createClient, http } from 'viem'
import { optimism } from 'tevm/common'
import { createTevmTransport } from 'tevm'

const client = createClient({
  transport: createTevmTransport({
    fork: { transport: http('https://mainnet.optimism.io')({}) }
  }),
  chain: optimism,
})

async function example() {
  // Get EOA (externally owned account) information
  const userAccount = await tevmGetAccount(client, {
    address: '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',
  })
  console.log(`Balance: ${userAccount.balance}`)
  console.log(`Nonce: ${userAccount.nonce}`)

  // Get contract account with storage
  const contractAccount = await tevmGetAccount(client, {
    address: '0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984', // UNI token
    returnStorage: true,
  })
  console.log(`Contract bytecode: ${contractAccount.code}`)
  console.log(`Storage slots: ${Object.keys(contractAccount.storage || {}).length}`)

  // Storage is a mapping of hex slot keys to hex values
  // Both are represented as hex strings
  if (contractAccount.storage) {
    for (const [slot, value] of Object.entries(contractAccount.storage)) {
      console.log(`Slot ${slot}: ${value}`)
    }
  }
}

example()
```

## See

 - [GetAccountParams](https://tevm.sh/reference/tevm/actions/type-aliases/getaccountparams/) for options reference.
 - [GetAccountResult](https://tevm.sh/reference/tevm/actions/type-aliases/getaccountresult/) for return values reference.
 - [TEVM Actions Guide](https://tevm.sh/learn/actions/)
 - [tevmSetAccount](https://tevm.sh/reference/tevm/actions/functions/setaccounthandler/) for modifying account state
