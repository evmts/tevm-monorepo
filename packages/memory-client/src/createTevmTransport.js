import { createCommon } from '@tevm/common'
import { requestEip1193, tevmSend } from '@tevm/decorators'
import { createTevmNode } from '@tevm/node'
import { createTransport } from 'viem'

/**
 * Creates a custom TEVM Transport for viem.
 *
 * A Transport in viem is the intermediary layer responsible for executing outgoing RPC requests. This custom TEVM Transport integrates an in-memory Ethereum client, making it ideal for local-first applications, optimistic updates, and advanced TEVM functionalities like scripting.
 *
 * @param {import('@tevm/node').TevmNodeOptions} options - Configuration options for the base client, similar to those used in `memoryClient` or a low-level `baseClient`.
 * @returns {import('./TevmTransport.js').TevmTransport} A configured TEVM transport.
 *
 * @example
 * ```typescript
 * import { createClient, http } from 'viem'
 * import { createTevmTransport } from 'tevm'
 * import { optimism } from 'tevm/common'
 *
 * const client = createClient({
 *   transport: createTevmTransport({
 *     fork: { transport: http('https://mainnet.optimism.io')({}) }
 *   }),
 *   chain: optimism,
 * })
 *
 * async function example() {
 *   const blockNumber = await client.getBlockNumber()
 *   console.log(blockNumber)
 * }
 *
 * example()
 * ```
 *
 * @see {@link createClient}
 * @see [Viem Client Docs](https://viem.sh/docs/clients/introduction)
 * @see [Client Guide](https://tevm.sh/learn/clients/)
 * @see [tevm JSON-RPC Guide](https://tevm.sh/learn/json-rpc/)
 * @see [EIP-1193 spec](https://eips.ethereum.org/EIPS/eip-1193)
 * @see [Ethereum jsonrpc docs](https://ethereum.org/en/developers/docs/apis/json-rpc/)
 * @see [CreateMemoryClient Docs](https://tevm.sh/reference/tevm/memory-client/functions/creatememoryclient/) - For a batteries-included client if not worried about tree shaking
 *
 * @typedef {Object} TevmTransportConfig
 * @property {number} [timeout=20000] - Timeout duration for requests in milliseconds. Default is 20,000 ms. Supplied by viem.
 * @property {number} [retryCount=3] - The maximum number of times to retry a failed request. Default is 3. Supplied by viem.
 * @property {import('viem').Chain} [chain] - Blockchain configuration. Defaults to the chain specified in `options` or the default TEVM chain.
 *
 * ## Parameters
 *
 * - `timeout` (optional, number): Timeout duration for requests in milliseconds. Default is 20,000 ms. Supplied by viem.
 * - `retryCount` (optional, number): The maximum number of times to retry a failed request. Default is 3. Supplied by viem.
 * - `chain` (optional, Chain): Blockchain configuration. Defaults to the chain specified in `options` or the default TEVM chain if not provided.
 *
 * ## Gotchas
 *
 * - When specifying a chain, use TEVM common instead of viem chains. You can create a TEVM common from a viem chain using `createCommon`.
 */
export const createTevmTransport = (options = {}) => {
	/**
	 * A map to store and manage TEVM clients keyed by chain ID.
	 * @type {Map<number, import('@tevm/node').TevmNode & import('@tevm/decorators').Eip1193RequestProvider & import('@tevm/decorators').TevmSendApi>}
	 */
	const tevmMap = new Map()

	/**
	 * Creates and returns a TEVM transport.
	 * @type {import('./TevmTransport.js').TevmTransport}
	 */
	return ({ timeout = 20_000, retryCount = 3, chain }) => {
		const dynamicChain =
			chain && 'ethjsCommon' in chain
				? /** @type {import('@tevm/common').Common} */ (chain)
				: chain !== undefined
					? // if user passed in chain instead of common create a common from it with cancun and default eips
						createCommon({ ...chain, hardfork: 'cancun', loggingLevel: 'warn' })
					: undefined
		const common = options.common ?? dynamicChain
		const id = common?.id ?? -1
		const tevm =
			tevmMap.get(id) ??
			createTevmNode({ ...options, ...(common !== undefined ? { common } : {}) })
				.extend(requestEip1193())
				.extend(tevmSend())
		tevmMap.set(id, tevm)

		return /** @type {any} */ (
			createTransport(
				{
					request: /** @type any */ (tevm.request),
					type: 'tevm',
					name: /** options?.name ?? */ 'Tevm transport',
					key: /* options?.key ?? */ 'tevm',
					timeout,
					retryCount,
					retryDelay: /* options?.retryDelay ?? */ 150,
				},
				{ tevm },
			)
		)
	}
}
