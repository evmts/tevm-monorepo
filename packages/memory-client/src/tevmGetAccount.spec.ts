import { optimism } from '@tevm/common'
import { transports } from '@tevm/test-utils'
import { type Client, createClient } from 'viem'
import { parseEther } from 'viem'
import { beforeEach, describe, expect, it } from 'vitest'
import type { TevmTransport } from './TevmTransport.js'
import { createTevmTransport } from './createTevmTransport.js'
import { tevmGetAccount } from './tevmGetAccount.js'

let client: Client<TevmTransport>
const prefundedAddress = '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'

beforeEach(async () => {
	client = createClient({
		transport: createTevmTransport({
			fork: { transport: transports.optimism },
		}),
		chain: optimism,
	})
})

describe('tevmGetAccount', () => {
	it('should get the account information for a prefunded address', async () => {
		const account = await tevmGetAccount(client, { address: prefundedAddress })
		expect(account).toBeDefined()
		expect(account.balance).toBe(parseEther('1000'))
		expect(account.nonce).toBe(0n)
		expect(account).toMatchSnapshot()
	})
})
