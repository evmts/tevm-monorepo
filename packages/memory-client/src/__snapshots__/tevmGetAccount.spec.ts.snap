// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tevmGetAccount > should get the account information for a prefunded address 1`] = `
{
  "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
  "balance": 1000000000000000000000n,
  "codeHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "deployedBytecode": "0x",
  "isContract": true,
  "isEmpty": false,
  "nonce": 0n,
  "storageRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
}
`;
